models/Amethyst/Prelude.agda

```diff
@@ -15,25 +15,92 @@
 --------------------------------------------------------------------------------
 module Amethyst.Prelude where
 
-open import Amethyst.Network public
-open import Amethyst.LinearAlgebra.As.Schmitty public
+open import Algebra.Core using (Op₁; Op₂)
+open import Data.Bool using (Bool)
+open import Data.Nat using (ℕ)
 open import Data.List as List public using (List; []; _∷_; _++_)
 open import Data.Vec as Vec public using (Vec; []; _∷_; [_])
+open import Data.Float using (Float)
+open import Relation.Binary.PropositionalEquality using (_≡_)
 open import SMT.Theories.Reals as Reals
+
+open import Amethyst.Network public
+open import Amethyst.LinearAlgebra.As.Schmitty public
+
 open import SMT.Backend.Z3 Reals.reflectable public using (z3)
 open import SMT.Backend.CVC4 Reals.reflectable public using (cvc4)
 
+--------------------------------------------------------------------------------
+-- Helper methods for generating constraints and queries
+
+NetworkConstraints : ℕ → ℕ → Set
+NetworkConstraints inputs outputs =
+  Vec (Real Γ) inputs →
+  Vec (Real Γ) outputs →
+  List (Term Γ BOOL)
+  where Γ = Reals inputs ++ Reals outputs
+
+processConstraints : ∀ {Γ} → List (Term Γ BOOL) → Command Γ [] []
+processConstraints constraints = assert (app₁ not (List.foldr (app₂ and) (app₀ true) constraints))
+
+query : ∀ {inputs outputs layers} (let Γ = Reals inputs ++ Reals outputs)
+  → Network Float inputs outputs layers
+  → NetworkConstraints inputs outputs
+  → Script [] Γ (SAT ∷ [])
```

Comment:
Is the let-binding really necessary? As far as I can tell, Γ is only used once.

Reply:
Yup, that's true it can be removed. At one point it was being repeated a lot.